### PR TITLE
add dbus commands for dynamic creation of instances

### PR DESCRIPTION
**Pull request by adriaan42**
Commit taken from #580

### CODE REVIEW

**Comment by adriaan42** [comment]
@yarda @zacikpa to continue the discussion on dynamic creation (and deletion) of plugin instances (originally in #580):

I have updated the behavior of `instance_destroy`:

* any instance can be destroyed (not only the ones created dynamically)
* devices attached to the instance are released, and later the plugin's `_add_device` function is called, and existing logic will attach the device to a suitable instance

For `instance_create` it's a bit more complex:

* I can extract `devices` and `devices_udev_regex` from the options and pass it to the instance creation. So when new devices are plugged, the new instance is taken into consideration.
* But what should happen to existing devices already attached to another instance?
  * Current implementation: nothing happens, and the user explicitly calls `instance_acquire_devices` to move devices, OR
  * Automatically move devices, i.e., iterate through all instances and check if any of their devices should be moved to the new instance. Would be possible with `_check_matching_devices`, while also considering instance priorities, but this would override any manual assignments made by previous calls to `instance_acquire_devices`... so not sure about the side effects.

Also, I noticed that the code (existing `instance_acquire_devices` and also my new `instance_create` and `instance_destroy`) call methods only available in Hotplug plugins, but without checking if we're actually dealing with a Hotplug plugin. Is that a bug in the existing `instance_acquire_devices`? Should I just add an `isinstance()` check to all three functions?

**Comment by adriaan42** [comment]
>   * Automatically move devices, i.e., iterate through all instances and check if any of their devices should be moved to the new instance. Would be possible with `_check_matching_devices`, while also considering instance priorities, but this would override any manual assignments made by previous calls to `instance_acquire_devices`... so not sure about the side effects.

After thinking about it some more, I believe that we do want to automatically move matching devices. So the latest update implements that, acquiring all matching devices from plugin instances with an equal or lower priority (i.e. if the priority value of the new instance is lower of equal to the one of the existing instance).

**Comment by adriaan42** [comment]
After some more testing and fixing, I'm now quite happy with the current implementation.

However, I'm seeing some potential inconsistencies in the current behavior of the `instance.active` property, where instances are only `active` when they have attached devices:

* non-active instances are not listed by the `get_instances` dbus call, so it seems they don't exist, but `instance_get_devices` can be called successfully -> the view exposed via the dbus interface is not consistent.
* for non-active instances, `instance_apply_tuning` (https://github.com/redhat-performance/tuned/blob/master/tuned/plugins/base.py#L250, and related functions) immediately returns, so in particular, the static tuning is not executed -> it is not possible to have static instance-specific tuning that is independent of devices.

**Review by zacikpa — CHANGES_REQUESTED** [review]
Thanks for the update, @adriaan42.

> Also, I noticed that the code (existing instance_acquire_devices and also my new instance_create and instance_destroy) call methods only available in Hotplug plugins, but without checking if we're actually dealing with a Hotplug plugin. Is that a bug in the existing instance_acquire_devices? Should I just add an isinstance() check to all three functions?

I think this is indeed a bug. It hasn't caused any issues yet because there aren't many non-hotplug plugins with device support, and for plugins that do not support devices at all, the existing function returns earlier. I would add the type check.

> non-active instances are not listed by the get_instances dbus call, so it seems they don't exist, but instance_get_devices can be called successfully -> the view exposed via the dbus interface is not consistent.

The only issue I see here is the name of the API function `get_instances`, which probably should have been named `get_active_instances`. Nonetheless, it's well-documented that it returns only active instances, so I wouldn't worry about it much.

> for non-active instances, instance_apply_tuning (https://github.com/redhat-performance/tuned/blob/master/tuned/plugins/base.py#L250, and related functions) immediately returns, so in particular, the static tuning is not executed -> it is not possible to have static instance-specific tuning that is independent of devices.

Good catch - I also found a related issue, see one of my comments. I think resolving this would require some heavier refactoring where we would decouple the non-device tuning from the device tuning. I don't really think it's a blocker for this PR - I would open a separate issue for that if you're not up for it. @yarda, what do you think?

Functionality-wise, I tested this and it works quite well.

File: tuned/daemon/controller.py [hunk 1]
```diff
@@ -417,3 +417,88 @@ def instance_get_devices(self, instance_name, caller = None):
...
+		Parameters:
+		plugin_name -- name of the plugin
+		instance_name -- name of the new instance
+		dict of string-string -- options for the new instance
+
+		Return:
+		bool -- True on success
+		string -- error message or "OK"
+		"""
+		plugins = {p.name: p for p in self._daemon._unit_manager.plugins}
```
[the remark this answers is not in the document]
Comment (adriaan42) [reply tuned/daemon/controller.py]: Yes, thanks! I've added those checks.

File: tuned/daemon/controller.py [hunk 2]
```diff
@@ -417,3 +417,88 @@ def instance_get_devices(self, instance_name, caller = None):
...
+			plugin.initialize_instance(instance)
+			self._daemon._unit_manager.instances.append(instance)
+		except Exception as e:
+			rets = "Error creating instance '%s': %s" % (instance_name, str(e))
+			log.error(rets)
+			return (False, rets)
+		log.info("Created dynamic instance '%s' of plugin '%s'" % (instance_name, plugin_name))
+
+		plugin.assign_free_devices(instance)
+		plugin.instance_apply_tuning(instance)
```
[the remark this answers is not in the document]
Comment (adriaan42) [reply tuned/daemon/controller.py:467]: Hmm... that would be a new feature, ensuring only one instance have active device-independent tuning?
In that case I'm not sure that this should be decided by priority (see also my comment https://github.com/redhat-performance/tuned/pull/580#issuecomment-1988809028).

For now I'm trying to give dynamically created instances the same behavior as "normal" ones.
And if I get the current implementation correctly, then if there are multiple active instances (defined in the profile), the device-independent tuning would run for all of them. And plugins wanting it differently need to implement something themselves (currently I think that only applies to `cpu`).

**Comment by adriaan42** [comment]
> Thanks for the update, @adriaan42.
> 
> > Also, I noticed that the code (existing instance_acquire_devices and also my new instance_create and instance_destroy) call methods only available in Hotplug plugins, but without checking if we're actually dealing with a Hotplug plugin. Is that a bug in the existing instance_acquire_devices? Should I just add an isinstance() check to all three functions?
> 
> I think this is indeed a bug. It hasn't caused any issues yet because there aren't many non-hotplug plugins with device support, and for plugins that do not support devices at all, the existing function returns earlier. I would add the type check.

I've added checks.

> > non-active instances are not listed by the get_instances dbus call, so it seems they don't exist, but instance_get_devices can be called successfully -> the view exposed via the dbus interface is not consistent.
> 
> The only issue I see here is the name of the API function `get_instances`, which probably should have been named `get_active_instances`. Nonetheless, it's well-documented that it returns only active instances, so I wouldn't worry about it much.

If inactive instance are "invisible" then any code using the interface to create a new instance would have to explicitly check (using `instance_get_devices`) if the chose name is already taken, or just try to create and catch the error. But I can live with that, and you're right, it's the documented behavior.

> > for non-active instances, instance_apply_tuning (https://github.com/redhat-performance/tuned/blob/master/tuned/plugins/base.py#L250, and related functions) immediately returns, so in particular, the static tuning is not executed -> it is not possible to have static instance-specific tuning that is independent of devices.
> 
> Good catch - I also found a related issue, see one of my comments. I think resolving this would require some heavier refactoring where we would decouple the non-device tuning from the device tuning. I don't really think it's a blocker for this PR - I would open a separate issue for that if you're not up for it. @yarda, what do you think?
> 
> Functionality-wise, I tested this and it works quite well.

Thanks, I appreciate it! :)

File: tuned/daemon/controller.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zacikpa) [reply tuned/daemon/controller.py:467]: Oh, you are right, this is not really related to the inconsistency. I guess we should define and document what we expect and not expect the `priority` option to do...

**Comment by adriaan42** [comment]
@yarda anything missing/blocking this?

**Comment by yarda** [comment]
> > for non-active instances, instance_apply_tuning (https://github.com/redhat-performance/tuned/blob/master/tuned/plugins/base.py#L250, and related functions) immediately returns, so in particular, the static tuning is not executed -> it is not possible to have static instance-specific tuning that is independent of devices.
> 
> Good catch - I also found a related issue, see one of my comments. I think resolving this would require some heavier refactoring where we would decouple the non-device tuning from the device tuning. I don't really think it's a blocker for this PR - I would open a separate issue for that if you're not up for it. @yarda, what do you think?
> 
OK, NP, it makes sense.

**Comment by yarda** [comment]
LGTM. I also tested basic functionality and it seems everything works OK, thanks for contribution.

**Pull request merged**